lib/Runtime/Language/JavascriptConversion.cpp

```diff
@@ -450,20 +450,21 @@ namespace Js
     //4. If SameValue(ToString(n), argument) is false, then return undefined.
     //5. Return n.
     //----------------------------------------------------------------------------
-    BOOL JavascriptConversion::CanonicalNumericIndexString(Var aValue, double *indexValue, ScriptContext * scriptContext)
+    BOOL JavascriptConversion::CanonicalNumericIndexString(JavascriptString *aValue, double *indexValue, ScriptContext * scriptContext)
     {
-        AssertMsg(JavascriptString::Is(aValue), "CanonicalNumericIndexString expects only string");
-        if (JavascriptString::IsNegZero(JavascriptString::FromVar(aValue)))
+        if (JavascriptString::IsNegZero(aValue))
         {
             *indexValue = -0;
             return TRUE;
         }
-        Var indexNumberValue = JavascriptOperators::ToNumber(aValue, scriptContext);
-        if (JavascriptString::Equals(JavascriptConversion::ToString(indexNumberValue, scriptContext), aValue))
+
+        double indexDoubleValue = aValue->ToDouble();
+        if (JavascriptString::Equals(JavascriptNumber::ToStringRadix10(indexDoubleValue, scriptContext), aValue))
```

Comment:
Only applied following optimizations on JavascriptConversion::CanonicalNumericIndexString():
- Type specialization (remove branches that are not applicable)
- Cancellation of boxing and unboxing (Number(Var) vs. double)
- Inlining
Equivalency is guaranteed.

---
In reply to: [106289593](https://github.com/Microsoft/ChakraCore/pull/2640#discussion_r106289593) [](ancestors = 106289593)